### ISSUE TITLE
No provider for AnimationEngine! is thrown after switch from 5.0.0-rc.1 to 5.0.0-rc.5

### ISSUE THREAD

**Issue by majo44** [angular/angular#19899]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x ] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior
The error:
```
ERROR { Error: StaticInjectorError[AnimationEngine]: 
  StaticInjectorError[AnimationEngine]: 
    NullInjectorError: No provider for AnimationEngine!
    at _NullInjector.get (...core.umd.js:962:19)
    at resolveToken (...\node_modules\@angular\core\bundles\core.umd.js:1250:24)
    at tryResolveToken (...\node_modules\@angular\core\bundles\core.umd.js:1192:16)
    at StaticInjector.get (...\node_modules\@angular\core\bundles\core.umd.js:1063:20)
    at resolveToken (...\node_modules\@angular\core\bundles\core.umd.js:1250:24)
    at tryResolveToken (...\node_modules\@angular\core\bundles\core.umd.js:1192:16)
    at StaticInjector.get (...\node_modules\@angular\core\bundles\core.umd.js:1063:20)
    at resolveNgModuleDep (...\node_modules\@angular\core\bundles\core.umd.js:10623:25)
    at _callFactory (...\node_modules\@angular\core\bundles\core.umd.js:10689:67)
    at _createProviderInstance$1 (...\node_modules\@angular\core\bundles\core.umd.js:10637:26)
``` 
is thrown on the server side rendering after switch from 5.0.0-rc.1 to 5.0.0-rc.5

## Expected behavior
The error should not be thrown 

## Minimal reproduction of the problem with instructions
Server Ng Module
```typescript
@NgModule({
    imports: [
        ServerModule,
        ServerTransferStateModule
        RouterModule.forRoot(ROUTES, { initialNavigation: 'enabled' }),
        AppModule,
    ],
    bootstrap: [AppComponent],
})
export class AppServerModule {
}
```

Common Ng Module
```typescript
@NgModule({
    imports: [
        CommonModule,
        FormsModule,
        RouterModule,
        BrowserModule.withServerTransition({appId: 'my-app'}),
    ],
    providers: [
        ...
    ],
    declarations: [...],
})
export class AppModule {
}
```
Rendering code

```typescript
import { Type } from '@angular/core';
import { renderModule } from '@angular/platform-server';
import * as express from 'express';
export function renderMiddleware(
    templateFilePath: string,
    module: Type<any>): express.RequestHandler {
    return async function middleware(request: express.Request, resonse: express.Response): Promise<void> {
        let result = await renderModule(module, {
            document: getDocument(templateFilePath),
            url: request.originalUrl,
        });
        resonse.send(result);
    };
}
```
## Environment

<pre><code>
Angular version: 5.0.0-rc.5

For Tooling issues:
- Node version: 6.11.0
- Platform:  Windows

Others:
</code></pre>

**Comment by vtabary** [comment]
For my project, the last working version is 5.0.0-rc.3. I tried the rc4, and it failed as well.

So something went wrong between the rc3 and the rc4

**Comment by patrickmichalina** [comment]
Same here https://github.com/patrickmichalina/fusebox-angular-universal-starter

**Comment by matsko** [comment]
@hansl you came across this error as well yes? @vikerman can you help figure out what's going on here? Usually this issue has been because of a version mismatch in `node_module` installs.

**Comment by patrickmichalina** [comment]
@matsko the error reminds me of similar DI issues from the DI change in 5.3 relating to static injection

**Comment by hansl** [comment]
@matsko Yes we have an issue filed against the CLI too.

@majo44 are you using the CLI? If so could you run `ng version`?

**Comment by majo44** [comment]
```
Angular CLI: 1.5.0-rc.3
Node: 6.11.0
OS: win32 x64
Angular: 5.0.0-rc.5
... animations, common, compiler, compiler-cli, core, forms
... http, platform-browser, platform-browser-dynamic
... platform-server, router

@angular/cli: 1.5.0-rc.3
@angular-devkit/build-optimizer: 0.0.30
@angular-devkit/core: 0.0.20
@angular-devkit/schematics: 0.0.34
@ngtools/json-schema: 1.1.0
@ngtools/webpack: 1.8.0-rc.3
@schematics/angular: 0.0.48
typescript: 2.5.3
webpack: 3.8.1
```

**Comment by tbosch** [comment]
This might be fixed on master already. So try again against rc.6.

**Comment by vikerman** [comment]
I cannot reproduce this issue. The universal-starter repo for CLI works without any issues.

@majo44 -  Can you share a repro github? 

The order of the modules is important. ServerModule should always be after AppModule

**Comment by vikerman** [comment]
@patrickmichalina  - Same for the fusebox starter. The ServerModule should appear after the AppModule.

**Comment by patrickmichalina** [comment]
@vikerman moving the module order did not fix this issue.

**Comment by vikerman** [comment]
Is it possible to share a github repro? I cannot reproduce this issue with our universal-starter

**Comment by coryrylan** [comment]
I ran into this on our universal Angular app running rc6 as well. I'll see if I can get anything. Unfortunately its a big internal app so will take some time to isolate.

**Comment by patrickmichalina** [comment]
@tbosch still broken in rc.6

**Comment by tbosch** [comment]
Hello,
could somebody create a branch on Github that reproduces the issue? Sorry, but we can't fix this otherwise...

**Comment by patrickmichalina** [comment]
@tbosch i can, but the starter requires a firebase account...

**Comment by alexzuza** [comment]
@patrickmichalina Is there a way to reproduce it without firebase?

**Comment by patrickmichalina** [comment]
@alexzuza  Ok. I made a branch that does not inject the firebase admin provider. And you can see the error no application start. This project does not use the angular-cli or webpack.

Upgraded this branch to rc.7

https://github.com/patrickmichalina/fusebox-angular-universal-starter/tree/provider-error

```
Error: StaticInjectorError[AnimationEngine]: 
  StaticInjectorError[AnimationEngine]: 
    NullInjectorError: No provider for AnimationEngine!
    at _NullInjector.get (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/di/injector.js:31:6)
    at resolveToken (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/@angular/core/bundles/core.umd.js:1250:24)
    at tryResolveToken (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/di/injector.js:330:14)
    at StaticInjector.get (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/di/injector.js:170:15)
    at resolveToken (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/@angular/core/bundles/core.umd.js:1250:24)
    at tryResolveToken (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/di/injector.js:330:14)
    at StaticInjector.get (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/di/injector.js:170:15)
    at resolveNgModuleDep (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/@angular/core/bundles/core.umd.js:10623:25)
    at _callFactory (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/view/ng_module.js:173:14)
    at _createProviderInstance$1 (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/view/ng_module.js:121:17)
    at resolveNgModuleDep (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/view/ng_module.js:103:2)
    at NgModuleRef_.get (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/@angular/core/bundles/core.umd.js:11844:16)
    at Object.createProdRootView [as createRootView] (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/@angular/core/bundles/core.umd.js:14063:62)
    at ComponentFactory_.create (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/view/refs.js:114:44)
    at ComponentFactoryBoundToModule.create (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/@angular/core/bundles/core.umd.js:3958:29)
    at ApplicationRef.bootstrap (/Users/Patrick/Repos/fusebox-angular-universal-starter/node_modules/packages/core/esm5/src/application_ref.js:686:55)
```

**Comment by tbosch** [comment]
@patrickmichalina I checked out your branch and tried `npm start --aot`, but that shows some errors in the command line, and only opens a browser that lists the directory. Could you provide more details how to reproduce?

Error I am seeing:
```
/Users/tbosch/projects/repro/issue-19899/fusebox-angular-universal-starter/node_modules/fuse-box-typechecker/dist/commonjs/src/checker.ts:171
        let allErrors = Object.entries(groupedErrors)
                               ^
TypeError: Object.entries is not a function
```

**Comment by tbosch** [comment]
Never mind, I used node 6, trying again...

**Comment by tbosch** [comment]
The error is gone, but still the app is not served. What are the right commands to run?

**Comment by patrickmichalina** [comment]
@tbosch I recommend just using `npm start`,

**Comment by alexzuza** [comment]
Here's my investigation. ServerModule
- imports `NoopAnimationsModule` that provides `AnimationEngine` 

```
import {ɵAnimationEngine as AnimationEngine, .....} from '@angular/animations/browser';

const SHARED_ANIMATION_PROVIDERS: Provider[] = [
  ...
  {provide: AnimationEngine, useClass: InjectableAnimationEngine}, 
  ...
];
```

- and also provides `SERVER_RENDER_PROVIDERS` that includes `RendererFactory2` token 

```
import {ɵAnimationEngine} from '@angular/animations/browser';
...
SERVER_RENDER_PROVIDERS  = [
 {
    provide: RendererFactory2,
    useFactory: instantiateServerRendererFactory,
    deps: [ServerRendererFactory2, ɵAnimationEngine, NgZone]
 },
```

As we can see `AnimationEngine` lives `'@angular/animations/browser'`. It looks nice, but let's take a look at generated bundles:

**platform-browser-animatons.umd.js** (it gives us `NoopAnimationsModule`)
```
define(...., '@angular/animations/browser') ....
  function(...._angular_animations_browser,...) {
.....

var SHARED_ANIMATION_PROVIDERS = [
  .....
  { provide: _angular_animations_browser.ɵAnimationEngine, useClass: InjectableAnimationEngine }
];
```
**platform-server.umd.js**

There is no import for `@angular/animations/browser` here but instead we see one more AnimationEngine.

```
var AnimationEngine = (function () {
    /**
     * @param {?} driver
     * @param {?} normalizer
     */
    function AnimationEngine(driver, normalizer) {

```
that becames our dependency for `RendererFactory2`

```
var SERVER_RENDER_PROVIDERS = [

    {
        provide: _angular_core.RendererFactory2,
        useFactory: instantiateServerRendererFactory,
        deps: [ServerRendererFactory2, AnimationEngine, _angular_core.NgZone]
    },
];

```

**Comment by tbosch** [comment]
@patrickmichalina Mmh, still same problem with `npm start`, i.e. it opens a browser for `http://localhost:8002/` that just shows the directory listing, and I can't find any `.html` files in there (seems to be the content of `node_modules/@angular`...)

**Comment by tbosch** [comment]
@alexzuza do you have a reproduction too? Could you paste it here?

**Comment by patrickmichalina** [comment]
@tbosch Wow I've never seen that happen before. By default it should open the browser to http://localhost:8000, but I've seen issues if another browsync process was already running, then a random port is assigned but won't connect.

**Comment by Toxicable** [comment]
@patrickmichalina Could you maybe delete out all of the unrelated stuff in there? there appears to be a lot of modules that are likely unrelated

**Comment by tbosch** [comment]
@patrickmichalina even if I open `http://localhost:8000/` I see the same problem (actually I can open `http://localhost:8000/`, `http://localhost:8001/`, and `http://localhost:8002/` with the same result).

**Comment by tbosch** [comment]
@patrickmichalina Sorry, I seem to have these ports open, let me change this.... Where can I change the port in your setup?

**Comment by alexzuza** [comment]
Yes, i can reproduce. 
https://unpkg.com/@angular/platform-server@5.0.0-rc.7/bundles/platform-server.umd.js

**Comment by Toxicable** [comment]
@alexzuza That's the source, we need a repo

**Comment by alexzuza** [comment]
I described why it's not working above based on the repo  https://github.com/patrickmichalina/fusebox-angular-universal-starter/tree/provider-error

**Comment by patrickmichalina** [comment]
@tbosch I recommend terminating other processes that may be conflicting with those ports or browsersync if possible. You can try to set via https://github.com/patrickmichalina/fusebox-angular-universal-starter/blob/provider-error/tools/config/build.config.ts.

Also I haven't tested in Windows so might be some funky build related stuff in there :(

**Comment by tbosch** [comment]
@patrickmichalina I think I was working off of master, and didn't use your branch `provider-error`.
@alexzuza you are right in your analysis, thanks! The question is why this changed between rc.3 and rc.4 (i.e. why we started to bundle `@angular/animations` into `platform-server` starting from there.

**Comment by alexzuza** [comment]
var VERSION = new _angular_core.Version('5.0.0-rc.7-5542517');

**Comment by tbosch** [comment]
@alexzuza ? what do you mean with this ?

**Comment by alexzuza** [comment]
`AnimationEngine` definition is still present in rc.8

https://unpkg.com/@angular/platform-server@5.0.0-rc.8/bundles/platform-server.umd.js

**Comment by vikerman** [comment]
Looks like the fix didn't make into rc.8

**Comment by hansl** [comment]
https://github.com/angular/angular/commits/5.0.x

It should have (look for `fix(platform-server): add missing packages to the UMD global rollup`). Maybe something else is missing?

**Comment by hansl** [comment]
@matsko See last 3 comments above ^. Could you give us a status on this? This is blocking the 5.0 release.

**Comment by vikerman** [comment]
Ah - sorry. Didn't see the merge on the PR. So the fix was merged into the 5.0 branch but looks like platform-server UMD bundle still has the AnimationEngine inlined :(

**Comment by alexzuza** [comment]
Something went wrong between rc3 https://unpkg.com/@angular/platform-server@5.0.0-rc.3/bundles/ and rc4 https://unpkg.com/@angular/platform-server@5.0.0-rc.4/bundles/

**Comment by mhevery** [comment]
Possible fix: https://github.com/angular/angular/issues/19899

**Comment by matsko** [comment]
Closed via 13f8648a002fb2a47bcf980eab5d087e54cafd74
